testbed/interfaces/node.go

```diff
@@ -0,0 +1,145 @@
+package testbedi
+
+import (
+	"context"
+	"io"
+)
+
+// NewNodeFunc constructs a node implementing the Core interface. It is provided
+// a path to an already created directory `dir`, as well as a map of attributes
+// which can be supplied to shape process execution.
+// Examples of attributes include: which binary to use, docker image, cpu/ram
+// limits, or any other information that may be required to property setup or
+// manage the node.
+type NewNodeFunc func(dir string, attrs map[string]interface{}) (Core, error)
+
+// GetAttrListFunc returns a list of attribute names that can be queried from
+// the node. These attributes may include those can be set from the NewNodeFunc,
+// or additional attributes at may only be available after initialization.
+// Attributes returned should be queriable through the Attribute interface.
+// Examples include: api address, peerid, cpu/ram limits, jitter.
+type GetAttrListFunc func() []string
+
+// GetAttrDescFunc returns the description of the attribute `attr`
+type GetAttrDescFunc func(attr string) (string, error)
+
+type Libp2p interface {
+	// PeerID returns the peer id
+	PeerID() (string, error)
+	// APIAddr returns the multiaddr for the api
+	APIAddr() (string, error)
+	// SwarmAddrs returns the swarm addrs for the node
+	SwarmAddrs() ([]string, error)
+}
+
+type Config interface {
+	Core
+	// GetConfig returns the configuration of the node
+	GetConfig() (interface{}, error)
+	// WriteConfig writes the configuration of the node
+	WriteConfig(interface{}) error
```

Comment:
Interfaces like this always make me a bit nervous. Be careful about sharing and race conditions. If this just writes a "json-like" thing, consider using refmt to copy it.

Reply:
Actually, what is the "form" of the config? I.e., what types of strings, what types of values, etc.

Reply:
I guess it's kind of completely up to the plugin to decide. Maybe these don't make a lot of sense anymore here. They would primarily be use when consuming iptb as a package though, allowing the user to get the config and cast it themselves.

https://github.com/ipfs/iptb/pull/61/files#diff-c348f01f34fe135d26d91d3176809188R106